crates/holochain/src/conductor/conductor/tests/agent_key_revocation.rs

```diff
@@ -0,0 +1,908 @@
+
+        // Cloning cells should fail for cell 1 as the agent key is invalid.
+        let mut create_clone_cell_payload = CreateCloneCellPayload {
+            role_name: role_1.to_string(),
+            membrane_proof: None,
+            modifiers: DnaModifiersOpt::none().with_network_seed("network_seed".into()),
+            name: None,
+        };
+        let result = conductor
+            .create_clone_cell(app.installed_app_id(), create_clone_cell_payload.clone())
+            .await
+            .unwrap_err();
+        assert_matches!(result, ConductorError::SourceChainError(SourceChainError::InvalidAgentKey(invalid_key, cell_id)) if invalid_key == agent_key && cell_id == cell_id_1);
+        // Cloning cells should succeed for cell 2 as the agent key is still valid.
+        create_clone_cell_payload.role_name = role_2.to_string();
+        let _ = conductor
+            .create_clone_cell(app.installed_app_id(), create_clone_cell_payload)
+            .await
+            .unwrap();
+
+        // Calling key revocation should succeed and return an error result for cell 1
+        let revocation_result_per_cell = conductor
+            .clone()
+            .revoke_agent_key_for_app(agent_key.clone(), app.installed_app_id().clone())
+            .await
+            .unwrap();
+        assert_matches!(revocation_result_per_cell.get(&cell_id_1).unwrap(), Err(ConductorApiError::SourceChainError(SourceChainError::InvalidAgentKey(key, cell_id))) if *key == agent_key && *cell_id == cell_id_1);
+        assert_matches!(revocation_result_per_cell.get(&cell_id_2).unwrap(), Ok(()));
+    }
+}
+
+mod multi_conductor {
+    use super::*;
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn revoke_agent_key_without_dpki() {
+        holochain_trace::test_run();
+        let no_dpki_conductor_config = SweetConductorConfig::rendezvous(true).no_dpki();
+        let mut conductors =
+            SweetConductorBatch::from_config_rendezvous(2, no_dpki_conductor_config).await;
+        let (dna_file, _, _) = SweetDnaFile::unique_from_test_wasms(vec![TestWasm::Create]).await;
+        let role = "role";
+        let apps = conductors
+            .setup_app("", [&(role.to_string(), dna_file.clone())])
+            .await
+            .unwrap();
+        let cells = apps.cells_flattened();
+        let alice = cells[0].agent_pubkey().clone();
+
+        await_consistency(20, &cells).await.unwrap();
+
+        // Deleting the key should succeed
+        let revocation_result_per_cell = conductors[0]
+            .clone()
+            .revoke_agent_key_for_app(alice.clone(), apps[0].installed_app_id().clone())
+            .await
+            .unwrap();
+        assert_matches!(
+            revocation_result_per_cell.get(cells[0].cell_id()).unwrap(),
+            Ok(())
+        );
+
+        await_consistency(20, &cells).await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn revoke_agent_key_with_dpki() {
```

Comment:
@jost-s unfortunately I think this test also needs to check the app cell, to ensure that Bob receives Alice's Delete action. That is the thing I'm concerned about and still not convinced that it will work, given that we invalidate the key in DPKI before committing the Delete action on the chain.

Reply:
Right, that's tested here https://github.com/holochain/holochain/pull/3930/files#diff-14062bb2839544a82ec826510201978b91f795dd579db231ca46127fe28d09e6R723